src/electionguard_gui/web/components/admin/upload-ballots-component.js

```diff
@@ -0,0 +1,122 @@
+import RouterService from "/services/router-service.js";
+import Spinner from "../shared/spinner-component.js";
+
+export default {
+  props: {
+    electionId: String,
+  },
+  components: { Spinner },
+  data() {
+    return {
+      election: null,
+      loading: false,
+      alert: null,
+      ballotsProcessed: null,
+      ballotsTotal: null,
+      success: false,
+    };
+  },
+  methods: {
+    async uploadBallots() {
+      try {
+        const form = document.getElementById("mainForm");
+        if (form.checkValidity()) {
+          this.loading = true;
+          this.alert = null;
+          this.ballotsProcessed = 0;
+          const ballotFiles = document.getElementById("ballotsFolder").files;
+          this.ballotsTotal = ballotFiles.length + 1;
+
+          const uploadId = await this.uploadDeviceFile();
+          await this.uploadBallotFiles(uploadId, ballotFiles);
+          this.success = true;
+        }
+        form.classList.add("was-validated");
+      } catch (ex) {
+        console.error(ex);
+        this.alert = ex.message;
+      } finally {
+        this.loading = false;
+      }
+    },
+    async uploadDeviceFile() {
+      const [deviceFile] = document.getElementById("deviceFile").files;
+      const deviceContents = await deviceFile.text();
+      console.log("Creating election", deviceFile.name);
+      const result = await eel.create_ballot_upload(
+        this.electionId,
+        deviceFile.name,
+        deviceContents
+      )();
+      if (!result.success) {
+        throw new Error(result.message);
+      }
+      this.ballotsProcessed++;
+      return result.result;
+    },
+    async uploadBallotFiles(uploadId, ballotFiles) {
+      for (let i = 0; i < ballotFiles.length; i++) {
+        const ballotFile = ballotFiles[i];
+        const ballotContents = await ballotFile.text();
```

Comment:
Reading in all the ballot files in JavaScript and sending the contents over web sockets to store in Mongo may not end up being the best approach since web sockets theoretically aren't as reliable as HTTP, but it seems to work well for me and it's all local traffic to the device.  If it ends up being a problem, we'll have to think pretty hard to find any better options.

Reply:
Since the ballot uploading is from the admin device and that is the one with the database on it, this should not be an issue at all.